torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,483 @@
+    """
+
+    def __init__(
+        self,
+        params,
+        optim: Type[Optimizer],
+        group: Optional[Any] = None,
+        bucket_cap_kb: int = 2 ** 24,
+        **default: Any,
+    ):
+        # Hold all the model params in the root .param_groups
+        self.initialized = False
+        super().__init__(params, default)
+
+        # Partition information. lazy evaluation, computed if requested
+        self._per_device_params: OrderedDict[
+            torch.device, List[List[Parameter]]
+        ] = OrderedDict()  # device, rank, params
+        self._param_rank: Dict[torch.Tensor, int] = {}
+        self._partition_parameters: List[List[Dict]] = []
+
+        # Build the wrapped optimizer, responsible for a shard of the params
+        self.group = group if group is not None else dist.group.WORLD
+        self.world_size = dist.get_world_size(self.group)
+        self.rank = dist.get_rank(self.group)
+        self.global_rank = _get_global_rank(self.group, self.rank)
+
+        self.optim = optim(self.partition_parameters()[self.rank], **default)
+
+        # - Sync local and global param_groups keys
+        for global_group, local_group in zip(self.param_groups, self.optim.param_groups):
+            for k, v in local_group.items():
+                if k != "params":
+                    global_group[k] = v
+
+        #  Optional consolidated optimizer state
+        self._all_states: List[Dict[str, Any]] = []
+
+        # Current default device is set by the parameters allocated to this rank
+        self._device = list(self.per_device_params.keys())[0]
+        self.buckets: Dict[torch.device, List[torch.Tensor]] = {}
+
+        self.bucket_size = bucket_cap_kb
+        for device, per_device in self.per_device_params.items():
+            # Allocate one buffer per rank and per device to group the small parameters
+            self.buckets[device] = [
+                torch.zeros(bucket_cap_kb, dtype=per_device[0][0].dtype, device=device) for _ in range(len(per_device))
+            ]
+        self.should_bucket_param: List[bool] = []
+        self.work_handles: Deque[Any] = deque()
+        self._setup_bucket_strategy()
+        self.initialized = True
+
+    def add_param_group(self, param_group: dict) -> None:
+        """Add a param group to the :class:`Optimizer` s `param_groups`.
+
+        This can be useful when fine tuning a pre-trained network as frozen layers can be made
+        trainable and added to the :class:`Optimizer` as training progresses.
+
+        Arguments:
+            param_group (dict): Specifies what Tensors should be optimized along with group
+            specific optimization options
+
+        .. warning: This handles updating the shards on all partitions, but needs to be called on all ranks.
+        """
+
+        super().add_param_group(param_group)
+        if self.initialized:
```

Comment:
curious, when will `self.initialized` be useful? Since `self.initialized` will always be `True` after ctor, do we expect concurrent calls to `add_param_group` when the optimizer didn't finish initialization?

Reply:
we call the parent constructor in ZeroRedundancyOptimizer's constructor, and the parent constructor in turn calls `add_param_group`. So it's not a matter of concurrency but inheritance, we'd like `add_param_group` to do something different depending on the call site

Reply:
Oh I see, shall we add some comments for that?